### PR TITLE
#699 #723  Category context, Category request data

### LINKED ISSUE #699: catalog.py:35-37: Create context class(es) for catalog...
The puzzle `683-588407ed` from #683 has to be resolved: 

https://github.com/fidals/shopelectro/blob/7f3bf67792297934c45486e6de8df62915076d32/shopelectro/views/catalog.py#L35-L37

The puzzle was created by Artemiy on 18-Jan-19. 

  

If you have any technical questions, don't ask me, submit new tickets instead. The task will be \"done\" when the problem is fixed and the text of the puzzle is _removed_ from the source code. Here is more about [PDD](http://www.yegor256.com/2009/03/04/pdd.html) and [about me](http://www.yegor256.com/2017/04/05/pdd-in-action.html).

### LINKED ISSUE #723: Wrap request args with some RequestData class
Now we have problem with request args. We calc this args in different not relevant places and pass them as one huge ungrouped args set. See code example below.

Incapsulate request args processing and passing with some universal `RequestData` class.

```
# it's part of shopelectro.views.catalog.CategoryPage.get_context_data method

        contexts = get_catalog_context(
            request=self.request,
            page=self.object,
            category=self.object.model,
            raw_tags=self.kwargs.get('tags'),
            page_number=int(self.request.GET.get('page', 1)),
            per_page=int(self.request.GET.get(
                'step', get_products_count(self.request),
            )),
            sorting_index=sorting_index,
        )
```

### PR DESCRIPTION
**Pull request by duker33**
Closes #699
Closes #723

- Move request processing from views to separated module (#723)
- Move context processing from views to separated module (#699)

### CODE REVIEW

**Comment by duker33** [comment]
@artemiy312 , sorry for such huge amount of code in the same chunk.
Please, log it's review with astronomical hours.

I made step throw PDD because we are in hurry with block stb task https://github.com/fidals/stroyprombeton/issues/431

File: shopelectro/request_data.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+import typing
+
+from django import http
+from django_user_agents.utils import get_user_agent
+
+
+class RequestData:
```
Comment (artemiy312) [inline shopelectro/request_data.py]:
```suggestion
class Request:
```

File: shopelectro/request_data.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    def __init__(
+        self, request: http.HttpRequest, url_kwargs: typing.Dict[str, str]
+    ):
+        self.request = request
+        self.url_kwargs = url_kwargs
+
+
+class ProductList(RequestData):
```
Comment (artemiy312) [inline shopelectro/request_data.py]:
```suggestion
class CatalogRequest(Request):
```

File: shopelectro/views/catalog.py [hunk 3]
```diff
@@ -210,69 +228,32 @@ def get_context_data(self, **kwargs):
...
-    :param slug: Slug for a given category
-    :param offset: used for slicing QuerySet.
-    :return: products list in html format
-    """
-    products_on_page = limit or get_products_count(request)
-    offset = int(offset)
-    if offset < 0:
+def load_more(request, **url_kwargs):
+    request_data_ = request_data.LoadMore(request, url_kwargs)
+    if request_data_.offset < 0:
```
Comment (artemiy312) [inline shopelectro/views/catalog.py]: Let's move this validation to `LoadMore`

File: shopelectro/views/catalog.py [hunk 4]
```diff
@@ -7,75 +7,93 @@
...
 from images.models import Image
-from pages import views as pages_views
 from shopelectro import models, context as se_context
+from shopelectro import request_data
 from shopelectro.views.helpers import set_csrf_cookie
 
-PRODUCTS_ON_PAGE_PC = 48
-PRODUCTS_ON_PAGE_MOB = 12
 
+class CatalogContext(newcontext.Context):
```
Comment (artemiy312) [inline shopelectro/views/catalog.py]: let's move this class to `shopelectro.context`

File: shopelectro/request_data.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        return int(self.url_kwargs.get('sorting', 0))
+
+    @property
+    def tags(self) -> str:
+        """Tags list in url args format."""
+        return self.url_kwargs.get('tags')
+
+    @property
+    def length(self):
+        """Max products list size based on device type."""
```
Comment (artemiy312) [inline shopelectro/request_data.py]:
```suggestion
        """Max size of products list depends on the device type."""
```

File: shopelectro/request_data.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    def slug(self):
+        return self.url_kwargs.get('slug', '')
+
+    @property
+    def sorting_index(self):
+        return int(self.url_kwargs.get('sorting', 0))
+
+    @property
+    def tags(self) -> str:
+        """Tags list in url args format."""
```
Comment (artemiy312) [inline shopelectro/request_data.py]: It ins't a clear comment. Let's remove it

File: shopelectro/request_data.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+class RequestData:
+    def __init__(
+        self, request: http.HttpRequest, url_kwargs: typing.Dict[str, str]
+    ):
+        self.request = request
+        self.url_kwargs = url_kwargs
+
+
+class ProductList(RequestData):
+    """Data came from django urls to django views."""
```
Comment (artemiy312) [inline shopelectro/request_data.py]:
```suggestion
    """Request came from django urls to django views."""
```

File: shopelectro/request_data.py (hunk 1 above)
Comment (artemiy312) [reply shopelectro/request_data.py]: We should avoid such anemic names

File: shopelectro/context.py [hunk 8]
```diff
@@ -25,3 +27,17 @@ def template_context(page, tag_titles, tags):
         return {
             'page': self._page,
         }
+
+
+class ListParams(Context):
```
Comment (artemiy312) [inline shopelectro/context.py]: Let's merge this class with `CatalogContext`

File: shopelectro/views/catalog.py (hunk 4 above)
Comment (artemiy312) [reply shopelectro/views/catalog.py]: Let's move `category`, `get_all_tags`, `filter_products` and `paginate_products` to `context` method as local variables.
They are zero-logic and almost declarative methods, which just make a noise and grow the size of `CatalogContext` in form of methods

File: shopelectro/request_data.py (hunk 2 above)
Comment (duker33) [reply shopelectro/request_data.py]: i'll change it to `Catalog`. It's already in `request_data` module

File: shopelectro/request_data.py (hunk 6 above)
Comment (duker33) [reply shopelectro/request_data.py]: changed it to `""":return: 'color=red&size=large' for example."""`

**Pull request merged**